contrib/babelfishpg_tsql/runtime/functions.c

```diff
@@ -5123,3 +5168,783 @@ get_bbf_pivot_tuplestore(const char 	*sourcetext,
+			xmlAttributePtr attr_def = NULL;
+
+			if (dtd != NULL)
+			{
+				/*
+				 * Its Unlikely that node->parent is NULL, Just a sanity check
+				 */
+				if (node->parent != NULL)
+					attr_def = xmlGetDtdAttrDesc(dtd, node->parent->name, node->name);
+
+				if (attr_def != NULL)
+				{
+					switch (attr_def->atype)
+					{
+						case XML_ATTRIBUTE_CDATA:
+							nulls[6] = false;
+							values[6] = PointerGetDatum((VarChar *) cstring_to_text("string"));
+							break;
+						case XML_ATTRIBUTE_ID:
+							nulls[6] = false;
+							values[6] = PointerGetDatum((VarChar *) cstring_to_text("id"));
+							break;
+						case XML_ATTRIBUTE_IDREF:
+							nulls[6] = false;
+							values[6] = PointerGetDatum((VarChar *) cstring_to_text("idref"));
+							break;
+						case XML_ATTRIBUTE_IDREFS:
+							nulls[6] = false;
+							values[6] = PointerGetDatum((VarChar *) cstring_to_text("idrefs"));
+							break;
+						case XML_ATTRIBUTE_ENTITY:
+							nulls[6] = false;
+							values[6] = PointerGetDatum((VarChar *) cstring_to_text("entity"));
+							break;
+						case XML_ATTRIBUTE_ENTITIES:
+							nulls[6] = false;
+							values[6] = PointerGetDatum((VarChar *) cstring_to_text("entities"));
+							break;
+						case XML_ATTRIBUTE_NMTOKEN:
+							nulls[6] = false;
+							values[6] = PointerGetDatum((VarChar *) cstring_to_text("nmtoken"));
+							break;
+						case XML_ATTRIBUTE_NMTOKENS:
+							nulls[6] = false;
+							values[6] = PointerGetDatum((VarChar *) cstring_to_text("nmtokens"));
+							break;
+						case XML_ATTRIBUTE_ENUMERATION:
+							nulls[6] = false;
+							values[6] = PointerGetDatum((VarChar *) cstring_to_text("enumeration"));
+							break;
+						case XML_ATTRIBUTE_NOTATION:
+							nulls[6] = false;
+							values[6] = PointerGetDatum((VarChar *) cstring_to_text("notation"));
+							break;
+						default:
+							break;
+					}
+				}				
+			}
+		}
+
+		/*
+		 * For attribute node prev and content are not applicable. So we should keep them NULL.
+		 */
+		if (node->type != XML_ATTRIBUTE_NODE)
+		{
+			if (node->prev != NULL)
+			{
+				xmlNodePtr cur = node->prev;
+
+				while (cur != NULL && cur->type == XML_TEXT_NODE && xmlIsBlankNode(cur))
+					cur = cur->prev;
+
+				node_id = lookup_xmlNode_id(cur);
+				if (node_id != -1)
+				{
+					nulls[7] = false;
+					values[7] = Int64GetDatum(node_id);
+				}
+			}
+
+			if (node->content != NULL)
+			{
+				char *ptr = (char *) node->content;
+
+				/* for content, trim leading and trailing spaces */
+				while (isspace((char) *ptr))
+					ptr++;
+
+				remove_trailing_spaces(ptr);
+
+				nulls[8] = false;
+				values[8] = PointerGetDatum(cstring_to_text((const char *) ptr));
+			}
+		}
+
+		if (!nulls[0] && !bms_is_member(DatumGetInt64(values[0]), *xml_visited_nodes_set))
+		{
+			tuplestore_putvalues(tupstore, tupdesc, values, nulls);
+			*xml_visited_nodes_set = bms_add_member(*xml_visited_nodes_set, DatumGetInt64(values[0]));
```

Comment:
what if node is already visited? I understand it might not be possible but should we have safeguard to return early before processing next?